torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -431,26 +433,40 @@ def powerSGD_hook(
             total_Qs_size, device=device, dtype=dtype
         )
 
+    # Batch tensors to compress by shape.
+    shape_to_tensors: Dict[torch.Size, List[torch.Tensor]] = {}
+    for tensor in tensors_to_compress:
+        if tensor.shape not in shape_to_tensors:
+            shape_to_tensors[tensor.shape] = []
+        shape_to_tensors[tensor.shape].append(tensor)
+
     # Create Ps and Qs that point to the allocated memory.
+    batched_tensors_to_compress = []
     ps = []
     qs = []
     p_idx = 0
     q_idx = 0
-    for tensor in tensors_to_compress:
-        n, m = tensor.shape
+    for shape, tensors in shape_to_tensors.items():
+        batch_size = len(tensors)
+        n, m = shape
         matrix_approximation_rank = min(n, m, state.matrix_approximation_rank)
+        if batch_size == 1:
+            # Use the original tensor to avoid copy.
+            batched_tensors_to_compress.append(tensors[0].unsqueeze(0))
+        else:
+            batched_tensors_to_compress.append(torch.stack(tensors))
```

Comment:
I guess will incur extra memory footprint and increase the peak memory usage. Shall we make such batching optimization configurable?

My concern is that for some memory-intensive use cases, enabling such optimization may cause OOM. Wdyt?